### PR TITLE
Removes and bans ImmutableSet

### PR DESCRIPTION
**Pull request by nik9000**
The next step in #13224.

### CODE REVIEW

**Comment by nik9000** [comment]
Note to any reviewers: be very very careful. I made a one line mistake and broke parent/child for a while during this process. The tests all pass but I'm a bit afraid.

Also, some of these changes might change the memory usage of the cluster state. ImmutableSet uses open addressing and a single array but this change switches to HashSets which resolves collisions with [trees](http://openjdk.java.net/jeps/180) - so more memory usage.

**Comment by nik9000** [comment]
In other odd news, this pr shaves a minute off of a clean compile and test of the core module. I'm not sure why.

File: core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java [hunk 1]
```diff
@@ -575,4 +600,12 @@ public static StreamInput wrap(byte[] bytes) {
...
         return new InputStreamStreamInput(new ByteArrayInputStream(bytes, offset, length));
     }
+
+    /**
+     * FunctionalInterface for methods that read an object from the stream.
+     */
+    @FunctionalInterface
+    public static interface ObjectReader<T> {
+        public T read(StreamInput in) throws IOException;
+    }
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java]: could this part of the PR be moved to another PR? Sorry I know this is a bit painful but since you asked for a careful review this would help me ensure everything is correct

File: core/src/main/java/org/elasticsearch/cluster/block/ClusterBlocks.java [hunk 2]
```diff
@@ -33,68 +33,68 @@
...
+    public Set<ClusterBlock> global(ClusterBlockLevel level) {
         return levelHolders[level.id()].global();
     }
 
-    public Map<String, ImmutableSet<ClusterBlock>> indices(ClusterBlockLevel level) {
+    public Map<String, Set<ClusterBlock>> indices(ClusterBlockLevel level) {
         return levelHolders[level.id()].indices();
     }
 
+    public Set<ClusterBlock> blocksForIndex(ClusterBlockLevel level, String index) {
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/cluster/block/ClusterBlocks.java]: can it be private?

File: core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java (hunk 1 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java]: Yeah - that is probably a good idea.

File: core/src/main/java/org/elasticsearch/cluster/block/ClusterBlocks.java (hunk 2 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/cluster/block/ClusterBlocks.java]: Yeah, I can switch that.

File: core/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java [hunk 3]
```diff
@@ -729,16 +743,16 @@ public static Builder builder(MetaData metaData) {
     }
 
     /** All known byte-sized cluster settings. */
-    public static final Set<String> CLUSTER_BYTES_SIZE_SETTINGS = ImmutableSet.of(
+    public static final Set<String> CLUSTER_BYTES_SIZE_SETTINGS = unmodifiableSet(newHashSet(
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java]: This can be left as a TODO for another PR but I don't see using Sets.newHashSet as an improvement, when we can directly use the JDK APIs and do `new HashSet<>(Arrays.asList(...))`

Comment (nik9000) [inline core/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java]: I suppose the difference here is that its _our_ `Sets.newHashSet` and not Guava's. I don't have a strong opinion here other than that we shouldn't remove that helper method in this PR. If we want it gone it can go in another one.

File: core/src/main/java/org/elasticsearch/common/util/set/Sets.java [hunk 4]
```diff
@@ -53,6 +59,27 @@ private Sets() {
...
+     * for initializing constants without static blocks.
+     */
+    public static <T> HashSet<T> newHashSetCopyWith(Set<T> original, T... elements) {
+        Objects.requireNonNull(original);
+        Objects.requireNonNull(elements);
+        HashSet<T> set = new HashSet<>(original.size() + elements.length);
+        set.addAll(original);
+        Collections.addAll(set, elements);
+        return set;
+    }
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/common/util/set/Sets.java]: I would rather have static blocks that this method?

File: core/src/main/java/org/elasticsearch/common/util/set/Sets.java [hunk 5]
```diff
@@ -53,6 +59,27 @@ private Sets() {
...
+
+    /**
+     * Collects {@code Stream<Collection<T>>} into {@code Set<T>}.
+     */
+    @SuppressWarnings("unchecked")
+    public static <T> Collector<Collection<T>, Set<T>, Set<T>> toFlatSet() {
+        return (Collector<Collection<T>, Set<T>, Set<T>>) (Object) ToFlatSetCollector.INSTANCE;
+    }
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/common/util/set/Sets.java]: this collector seems to have only one consumer, maybe we don't really need it?

File: core/src/main/java/org/elasticsearch/script/ScriptContextRegistry.java [hunk 6]
```diff
@@ -76,15 +81,11 @@ private void validateScriptContext(ScriptContext.Plugin scriptContext) {
...
-            builder.add(scriptType.toString());
-        }
-        for (ScriptContext.Standard scriptContext : ScriptContext.Standard.values()) {
-            builder.add(scriptContext.getKey());
-        }
-        builder.add("script").add("engine");
-        return builder.build();
+    private static Set<String> reservedScriptContexts() {
+        Set<String> reserved = concat(stream(ScriptService.ScriptType.values()), stream(ScriptContext.Standard.values()))
+                .map(Object::toString).collect(toSet());
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/script/ScriptContextRegistry.java]: hmm, we need to call `getKey`, not `toString` on the elements of the second array?

**Comment by jpountz** [comment]
It looks good to me overall. I think the fact that the PR both moves to the stream API and forbids Immutable makes the review a bit harder, for future refactorings maybe we should try to do it in different pull requests. As far as I am concerned, I think I would find the use of the stream API a bit more readable if the chaining was performed on a separate line for every operation.

File: core/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java (hunk 3 above)
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java]: > we shouldn't remove that helper method in this PR

ohhh yeaaah, that's fore sure!

File: core/src/main/java/org/elasticsearch/common/util/set/Sets.java (hunk 4 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/util/set/Sets.java]: I can pitch it.

File: core/src/main/java/org/elasticsearch/common/util/set/Sets.java (hunk 5 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/util/set/Sets.java]: I'll remove it.

File: core/src/main/java/org/elasticsearch/script/ScriptContextRegistry.java (hunk 6 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/script/ScriptContextRegistry.java]: Fixed.

**Comment by nik9000** [comment]
> I think the fact that the PR both moves to the stream API and forbids Immutable makes the review a bit harder, for future refactorings maybe we should try to do it in different pull requests.

I imagine we wouldn't really move to Stream at all if we did that - which is ok I guess.

> I think I would find the use of the stream API a bit more readable if the chaining was performed on a separate line for every operation.

I've done that a few places. I also reverted back to imperative style in a few places as well.

File: core/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java (hunk 3 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java]: I know this is kind of a burried place to have the discussion, but what is wrong with `newHashSet` as a utility? That people will use it and not see the implicit array creation?

**Comment by nik9000** [comment]
@jpountz I believe this is ready for round two at your leisure. Thanks for reviewing it!

File: core/src/main/java/org/elasticsearch/common/inject/MembersInjectorImpl.java [hunk 7]
```diff
@@ -112,11 +115,7 @@ public String toString() {
...
     }
 
-    public ImmutableSet<InjectionPoint> getInjectionPoints() {
-        ImmutableSet.Builder<InjectionPoint> builder = ImmutableSet.builder();
-        for (SingleMemberInjector memberInjector : memberInjectors) {
-            builder.add(memberInjector.getInjectionPoint());
-        }
-        return builder.build();
+    public Set<InjectionPoint> getInjectionPoints() {
+        return unmodifiableSet(memberInjectors.stream().map(SingleMemberInjector::getInjectionPoint).collect(toSet()));
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/common/inject/MembersInjectorImpl.java]: one operation per line would be more readable?

Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/inject/MembersInjectorImpl.java]: Ahk. Unpushed local changes.

**Comment by nik9000** [comment]
Sorry about that! I had unpushed local changes that had addressed some of the formatting.

**Comment by jpountz** [comment]
LGTM

**Comment by nik9000** [comment]
I've merged master into this to pick up the query refactoring changes that were just merged.
